packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3873,3 +3880,243 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller.removeListener(_push);
+      widget.controller.addListener(_push);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller.removeListener(_push);
+    _throttleTimer?.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Actions(
+      actions: <Type, Action<Intent>> {
+        UndoTextIntent: CallbackAction<UndoTextIntent>(onInvoke: _undo),
+        RedoTextIntent: CallbackAction<RedoTextIntent>(onInvoke: _redo),
+      },
+      child: widget.child,
+    );
+  }
+}
+
+/// A data structure representing a chronological list of states that can be
+/// undone and redone.
+@visibleForTesting
+class UndoStack<T> {
+  /// Creates an instance of [UndoStack].
+  UndoStack();
+
+  final List<T> _list = <T>[];
+
+  // The index of the current value, or null if the list is emtpy.
+  late int _index;
+
+  /// Returns the current value of the stack.
+  T? get currentValue => _list.isEmpty ? null : _list[_index];
+
+  /// Add a new state change to the stack.
+  ///
+  /// Pushing identical objects will not create multiple entries.
+  void push(T value) {
+    if (_list.isEmpty) {
+      _index = 0;
+      _list.add(value);
+      return;
+    }
+
+    assert(_index < _list.length && _index >= 0);
+
+    if (value == currentValue) {
+      return;
+    }
+
+    // If anything has been undone in this stack, remove those irrelevant states
+    // before adding the new one.
+    if (_index != null && _index != _list.length - 1) {
+      _list.removeRange(_index + 1, _list.length);
+    }
+    _list.add(value);
+    _index = _list.length - 1;
+  }
+
+  /// Returns the previous state.
+  ///
+  /// If the _list is empty, returns null.
+  T? undo() {
+    if (_list.isEmpty) {
+      return null;
+    }
+
+    assert(_index < _list.length && _index >= 0);
+
+    if (_index != 0) {
+      _index = _index - 1;
+    }
+
+    return _list[_index];
+  }
+
+  /// Returns the state that was last undone, or if none, the current state.
+  ///
+  /// If the _list is empty, returns null.
+  T? redo() {
+    if (_list.isEmpty) {
+      return null;
+    }
+
+    assert(_index < _list.length && _index >= 0);
+
+    if (_index < _list.length - 1) {
+      _index = _index + 1;
+    }
+
+    return _list[_index];
+  }
+
+  @override
+  String toString() {
+    return 'UndoStack $_list';
+  }
+}
+
+/// A function that can be, or has been, throttled with the throttle function.
+@visibleForTesting
+typedef Throttleable<T> = void Function(T currentArg);
+
+/// A function that has been throttled by [throttle].
+@visibleForTesting
+typedef Throttled<T> = Timer Function(T currentArg);
+
+/// Returns a _Throttled that will call through to the given function only a
+/// maximum of once per duration.
+///
+/// Only works for functions that take exactly one argument.
```

Comment:
It doesn't matter for UndoRedo, but I couldn't get this to work with more than one argument because it doesn't seem possible to refer to all arguments in Dart.  Maybe Dart isn't meant for this kind of functional programming.